src/storage/src/table/state_table.rs

```diff
@@ -102,28 +111,128 @@ impl<S: StateStore> StateTable<S> {
         Ok(())
     }
 
-    pub async fn iter(&self, _pk: Row) -> StorageResult<StateTableRowIter<S>> {
-        todo!()
+    pub async fn iter(&self, epoch: u64) -> StorageResult<StateTableRowIter<'_, S>> {
+        let mem_table_iter = self.mem_table.buffer.iter().peekable();
+        StateTableRowIter::new(
+            &self.keyspace,
+            &self.column_descs,
+            mem_table_iter,
+            &self.order_types,
+            epoch,
+        )
+        .await
     }
 }
 
-pub struct StateTableRowIter<S: StateStore> {
-    cell_based_iter: CellBasedTableRowIter<S>,
-    mem_table: MemTable,
+/// `StateTableRowIter` is able to read the just written data (uncommited data).
+/// It will merge the result of `mem_table_iter` and `cell_based_streaming_iter`
+pub struct StateTableRowIter<'a, S: StateStore> {
+    mem_table_iter: Peekable<MemTableIter<'a>>,
+    cell_based_streaming_iter: CellBasedTableStreamingIter<S>,
+    /// The result of the last cell_based_streaming_iter next is saved, which can avoid being
+    /// discarded.
+    pk_serializer: OrderedRowSerializer,
 }
+type MemTableIter<'a> = Iter<'a, Row, RowOp>;
 
-impl<S: StateStore> StateTableRowIter<S> {
+impl<'a, S: StateStore> StateTableRowIter<'a, S> {
     async fn new(
-        _keyspace: Keyspace<S>,
-        _table_descs: Vec<ColumnDesc>,
-        _epoch: u64,
-        _stats: Arc<StateStoreMetrics>,
-    ) -> StorageResult<Self> {
-        todo!()
+        keyspace: &Keyspace<S>,
+        table_descs: &[ColumnDesc],
+        mem_table_iter: Peekable<MemTableIter<'a>>,
+        order_types_vec: &[OrderType],
+        epoch: u64,
+    ) -> StorageResult<StateTableRowIter<'a, S>> {
+        let cell_based_streaming_iter =
+            CellBasedTableStreamingIter::new(keyspace, table_descs, epoch).await?;
+        let pk_serializer = OrderedRowSerializer::new(order_types_vec.to_vec());
+        let state_table_iter = Self {
+            mem_table_iter,
+            cell_based_streaming_iter,
+            pk_serializer,
+        };
+
+        Ok(state_table_iter)
     }
 
-    async fn next(&mut self) -> StorageResult<Option<Row>> {
-        todo!()
+    pub async fn next(&mut self) -> StorageResult<Option<Row>> {
+        // todo(wcy-fdu): use async_stream to implement this part.
+        self.cell_based_streaming_iter
+            .init_cell_based_item()
+            .await
+            .map_err(err)?;
+
+        let mut mem_table_iter_next_flag = false;
+        let mut res = None;
+        match (
+            self.cell_based_streaming_iter.cell_based_item.as_ref(),
+            self.mem_table_iter.peek(),
+        ) {
+            (None, None) => {
+                res = None;
+            }
+            (Some((_, row)), None) => {
+                res = Some(row.clone());
```

Comment:
We should avoid this `clone` since we are going to call `next` of cell-based streaming iterator.

Reply:
I suggest that you get the owned value from next instead. Currently, the clone is performed at least twice for cell based. The flow should be:
```
peek -> if stored is None, call next() and cache as Some(_).
next -> If Some(_), return, set to None. If None, perform work to get next value.
```

Reply:
Can just make everything easier and wrap in Peekable: https://doc.rust-lang.org/stable/src/core/iter/adapters/peekable.rs.html#15-19